Add pipes to CPK funding and redeem

hexyls opened the pull request:
related #2031 

https://github.com/protofire/omen-exchange/pull/2074 should be merged first 

Testing:

- [ ] Add funding to categorical xdai market
- [ ] Remove funding from scalar xdai market
- [ ] Add funding to categorical eth market
- [ ] Remove funding from scalar eth market
- [ ] Can bond + redeem on a market

KadenZipfel commented:
Testing:

- [ ] Add funding to categorical xdai market
- [ ] Remove funding from scalar xdai market
- [ ] Add funding to categorical eth market
- [ ] Remove funding from scalar eth market
- [ ] Can bond + redeem on a market

Mi-Lan commented:
- [ ] Add funding to categorical xdai market
- [ ] Remove funding from scalar xdai market
- [ ] Add funding to categorical eth market
- [ ] Remove funding from scalar eth market
- [ ] Can bond + redeem on a market

Mi-Lan commented:
I got the error the same as kaden but transaction went through...https://rinkeby.etherscan.io/tx/0xb205ee7b6522146e2ebd221f9bced0e6e60b28ff7c08b7dc5fde59e1c6a245ce
Heres video as well....
https://user-images.githubusercontent.com/33226956/124968131-8d104e80-e025-11eb-85bc-bf514aab2a1e.mp4

hexyls commented:
@KadenZipfel @Mi-Lan I tried some different market configs but couldn't reproduce the redeem error 😩 could you please document what market type, collateral, bets made, how you bonded / redeemed etc so I can hunt down this bug? Cheers guys

pimato commented:
Testing:

- [ ] Add funding to categorical xdai market
- [ ] Remove funding from scalar xdai market
- [ ] Add funding to categorical eth market
- [ ] Remove funding from scalar eth market
- [ ] Can bond + redeem on a market

Mi-Lan commented:
> @KadenZipfel @Mi-Lan I tried some different market configs but couldn't reproduce the redeem error 😩 could you please document what market type, collateral, bets made, how you bonded / redeemed etc so I can hunt down this bug? Cheers guys

Interesting 🤔  I created scalar eth market, then deposited some liqudity,then bought some long shares and then bet on outcome that had my long shares win out...I had also short shares from providing liquditiy...I will try to reproduce it again

KadenZipfel commented:
> @KadenZipfel @Mi-Lan I tried some different market configs but couldn't reproduce the redeem error 😩 could you please document what market type, collateral, bets made, how you bonded / redeemed etc so I can hunt down this bug? Cheers guys

Steps to reproduce:
- Go to pending markets on rinkeby
- Bond on an outcome of an eth market
- Try to redeem

hexyls commented:
I think @KadenZipfel discovered a super weird issue, it does not appear to be directly related to this PR but can be prevented by adding a function to the CPK market creation pipeline, seems like there are two problems:

#### oracle address mismatch

`conditionId` is determined via oracle address, questionId, and outcomeSlotCount, it looks like we updated the rinkeby oracle address last year and because we use the hardcoded oracle address across all markets it breaks `conditionId` calculations for older markets (so you cannot resolve them).

I updated this PR to make sure we are using the indexed oracle address for each individual market rather than the hardcoded one when resolving a market. 


#### conditional tokens or realitio address mismatch
Even with the change above some older markets are fkd lol it is possible to put a market into a state where you cannot resolve it. e.g. [this market is in this state](http://localhost:3000/#/0x26b5321d2b0fbd19482fab35afa054af036e5db6/finalize) - This can happen when we update the hardcoded conditionalTokens or realitio contract addresses without deploying a new oracle contract.

The market linked above had [`prepareCondition`](https://dashboard.tenderly.co/tx/rinkeby/0x908a67bc33180491aa576b39f78d5aa3f622eb9538e3f60527353f511117c997) called on a new conditionalTokens contract while the [oracle points to an older conditionalTokens contract](https://rinkeby.etherscan.io/address/0x576b76eebe6b5411c0ef310e65de9bff8a60130f#readContract), so when you attempt to resolve the market there is an address mismatch. I don't think there is anything we can do for older markets, but I have added an oracle validation into the market creation pipeline to prevent this in the future.

This will still remain as somewhat of a security issue - because the oracle is not validated on-chain a malicious user can deploy markets that cannot be resolved.

KadenZipfel commented:
> This will still remain as somewhat of a security issue - because the oracle is not validated on-chain a malicious user can deploy markets that cannot be resolved.

@hexyls can we incorporate something on the front-end to say that a market has this issue to prevent any malicious activity?

hexyls commented:
@KadenZipfel yeah sounds good, opened a new issue: https://github.com/protofire/omen-exchange/issues/2099

Ok so now that we know about that - let's try testing this PR again with only new markets

KadenZipfel commented:
Testing on new markets:

- [x] Add funding to categorical xdai market
- [x] Remove funding from scalar xdai market
- [x] Add funding to categorical eth market
- [x] Remove funding from scalar eth market
- [x] Can bond + redeem on a market

KadenZipfel commented on app/src/services/cpk/fns.ts:
```diff
@@ -345,15 +584,117 @@ export const prepareCondition = async (params: PrepareConditionParams) => {
+export const claimWinnings = async (params: ClaimWinningsParams) => {
+  const { isConditionResolved, question, realitio, transactions } = params
+
+  if (!isConditionResolved) {
```
Shouldn't this be `if (isConditionResolved)`?

hexyls replied:
`resolveCondition` is the first function in the pipeline so by the time `claimWinnings` is called the condition is resolved + it is super weird but `claimWinnings` is only called once and it claims winnings for all bonders into their realitio balance, `withdrawRealitioBalance` withdraws individual winnings

Mi-Lan commented:
Testing:

- [ ] Add funding to categorical xdai market
- [ ] Remove funding from scalar xdai market
- [ ] Add funding to categorical eth market
- [ ] Remove funding from scalar eth market
- [ ] Can bond + redeem on a market

Mi-Lan reviewed (approved):
LGTM!

pimato commented:
Testing:

- [x] Add funding to categorical xdai market

- [x] Remove funding from scalar xdai market

- [x] Add funding to categorical eth market

- [x] Remove funding from scalar eth market

- [x] Can bond + redeem on a market
- Bond

- Redeem

pimato reviewed (approved):
works great!
